src/Analyser/NodeScopeResolver.php

```diff
@@ -139,1146 +188,2810 @@ public function setAnalysedFiles(array $files)
+
+			$resultType = TypeCombinator::removeNull($leftType)->combineWith($rightType);
+			return $leftScope->mergeWith($rightScope);
+		}
+
+		$scope = $this->processExprNode($node->left, $scope, $nodeCallback, $leftType);
+		$scope = $this->processExprNode($node->right, $scope, $nodeCallback, $rightType);
+
+		if ($leftType instanceof BooleanType) {
+			$leftType = new IntegerType();
+		}
+
+		if ($rightType instanceof BooleanType) {
+			$rightType = new IntegerType();
+		}
+
+		if ($node instanceof BinaryOp\Plus) {
+			if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+				$resultType = new IntegerType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof IntegerType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof IntegerType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof ArrayType && $rightType instanceof ArrayType) {
+				$resultType = new ArrayType($leftType->getItemType()->combineWith($rightType->getItemType()));
+
+			} else {
+				$resultType = new MixedType();
+			}
+
+		} elseif ($node instanceof BinaryOp\Minus) {
+			if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+				$resultType = new IntegerType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof IntegerType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof IntegerType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} else {
+				$resultType = new MixedType();
+			}
+
+		} elseif ($node instanceof BinaryOp\Mul) {
+			if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+				$resultType = new IntegerType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof IntegerType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof IntegerType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} else {
+				$resultType = new MixedType();
+			}
+
+		} elseif ($node instanceof BinaryOp\Div) {
+			if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+				$resultType = new FloatType(); // TODO: union
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof IntegerType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof IntegerType && $rightType instanceof FloatType) {
+				$resultType = new FloatType();
+
+			} elseif ($leftType instanceof FloatType && $rightType instanceof FloatType) {
+				
```

Comment:
instead of this if-else, there should probably be
```php
$ifTrueScope = $this->processExprNode($node->if ?? $node->cond, $ifTrueScope, $nodeCallback, $ifTrueResultType);
$ifFalseScope = $this->processExprNode($node->else, $ifFalseScope, $nodeCallback, $ifFalseResultType);
```